### ISSUE TITLE
Reporting version info for maven dependencies

### ISSUE THREAD

**Issue by jtmelton** [dependency-check/DependencyCheck#605]
This is an enhancement idea specifically around the maven mojo - I'm willing to take a first cut at this work. Looking for feedback on a) would it be useful and b) discussion of solution.

First question, would it be useful to be able to present the available versions? I see there's a ```getVersionEvidence()``` method that exists on the ```Dependency``` class, but don't see it used in the system. Would it be useful to add this data for maven?

If the answer above is "yes", I have some follow-up questions: 
- would you like to see _all_ available versions or just _future_ available versions, ie. if _all_ is [1.0, 2.0, 3.0, 4.0, 5.0] and _current_ is 3.0, then _future_ would be [4.0, 5.0]. (I think the "right" answer is complex here. It could be absolute latest. It could be any non-vulnerable version ... assuming there's a CVE. It could be latest available on my major release, assuming semantic versioning).
- what would you like to see as the name (string) of the available version. Do you want just the version number itself or the full GAV?

**Comment by jeremylong** [comment]
Great suggestion - yes the versions that one could upgrade to would be valuable (heck we might be able to see if they are known vulnerable too).  Would this have to be Maven specific though? If the hash matches (either CentralAnalyzer or NexusAnalyzer) isn't there a subsequent get available versions API?

Regarding the `getVersionEvidence()` - that is the evidence used for identification of the correct version within the CVE/CPE data. It may contain the GAV version, but it would not be used for available versions. We would have to extend the dependency object to contain available version information.

Regarding what to present - I agree this is difficult. I might consider taking the exact same approach as the dependency plugin and show the latest minor and latest major release (we might even be able to borrow code depending on the licenses for this).  One of the known issues with the get available versions is that sometimes there is a timestamp or something in the version that completely messes up trying to figure out what the latest is.

For reporting purposes - the full GAV of the available version would be good.

**Comment by jtmelton** [comment]
Not very familiar with nexus/central apis. Probably good to loop in @colezlaw on this one as it appears he originally did that work.

That sounds good to extend the object to hold this info. Are any of the other analyzers doing something similar or is this new project-wide?

Yes, agree the get available versions is non-optimal as not everyone uses semantic versioning. I'll try to take a peek at the dependency plugin once we hash out the other issues.

**Comment by jtmelton** [comment]
Bumping this to @colezlaw to see if he has any thoughts.

**Comment by jeremylong** [comment]
Baked into Maven is an API to get the available versions. The problem is this could probably work for Maven, gradle, or sbt - it likely would not be available in the other integrations (unless we found the maven coordinates from within the JAR in the CLI/ANT version).

**Comment by jtmelton** [comment]
Yeah, I think this would definitely be just for those integrations where we have access to the maven coordinates easily, at least at first. I was using those apis for the POC.
